File: 1/2021.03.26/654.Maximum Binary Tree/python/solution.py
class Solution:
    def constructMaximumBinaryTree(self, nums: List[int]) -> TreeNode:
        def build(start, end):
            if start > end:
                return None
            # find max
            max_pos = start
            for i in range(start + 1, end + 1):
                if nums[i] > nums[max_pos]:
                    max_pos = i
            root = TreeNode(nums[max_pos])
            root.left = build(start, max_pos - 1)
            root.right = build(max_pos + 1, end)
            return root
        return build(0, len(nums) - 1)

File: 1/2021.03.26/654.Maximum Binary Tree/java/solution.java
class Solution {
    public TreeNode constructMaximumBinaryTree(int[] nums) {
        return build(nums, 0, nums.length - 1);
    }

    private TreeNode build(int[] nums, int start, int end) {
        if (start > end) {
            return null;
        }
        int maxPos = start;
        for (int i = start + 1; i <= end; i++) {
            if (nums[i] > nums[maxPos]) {
                maxPos = i;
            }
        }
        TreeNode root = new TreeNode(nums[maxPos]);
        root.left = build(nums, start, maxPos - 1);
        root.right = build(nums, maxPos + 1, end);
        return root;
    }
}
